lib/Monads/trace/Trace_Prefix_Closed.thy

```diff
@@ -0,0 +1,301 @@
+(*
+ * Copyright 2023, Proofcraft Pty Ltd
+ * Copyright 2020, Data61, CSIRO (ABN 41 687 119 230)
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ *)
+
+theory Trace_Prefix_Closed
+  imports
+    Trace_No_Trace
+    Trace_Monad_Equations
+    Eisbach_Tools.Rule_By_Method
+begin
+
+subsection "Prefix Closed"
+
+definition prefix_closed :: "('s, 'a) tmonad \<Rightarrow> bool" where
+  "prefix_closed f = (\<forall>s. \<forall>x xs. (x # xs) \<in> fst ` f s \<longrightarrow> (xs, Incomplete) \<in> f s)"
+
+lemmas prefix_closedD1 = prefix_closed_def[THEN iffD1, rule_format]
+
+lemma in_fst_snd_image_eq:
+  "x \<in> fst ` S = (\<exists>y. (x, y) \<in> S)"
+  "y \<in> snd ` S = (\<exists>x. (x, y) \<in> S)"
+  by (auto elim: image_eqI[rotated])
+
+lemma in_fst_snd_image:
+  "(x, y) \<in> S \<Longrightarrow> x \<in> fst ` S"
+  "(x, y) \<in> S \<Longrightarrow> y \<in> snd ` S"
+  by (auto simp: in_fst_snd_image_eq)
+
+lemmas prefix_closedD = prefix_closedD1[OF _ in_fst_snd_image(1)]
+
+lemma no_trace_prefix_closed:
+  "no_trace f \<Longrightarrow> prefix_closed f"
+  by (auto simp add: prefix_closed_def dest: no_trace_emp)
+
+
+subsection \<open>Set up for @{method wp}\<close>
+
+lemma prefix_closed_is_triple[wp_trip]:
+  "prefix_closed f = triple_judgement \<top> f (\<lambda>() f. id prefix_closed f)"
+  by (simp add: triple_judgement_def split: unit.split)
+
+
+subsection \<open>Rules\<close>
+
+(*
+  Collect generic prefix_closed lemmas here:
+   - naming convention is prefix_closed_NAME.
+   - add lemmas with assumptions to [prefix_closed_cond] set
+   - add lemmas without assumption to [prefix_closed_terminal] set
+*)
+
+named_theorems prefix_closed_terminal
+named_theorems prefix_closed_cond
+
+lemmas [prefix_closed_terminal] = no_trace_terminal[THEN no_trace_prefix_closed]
```

Comment:
Neat, that saves a lot of typing.

Reply:
Yeah, I was pretty happy that this worked.